### PR TITLE
Changed api of LinkGenerator

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing.Abstractions/LinkOptions.cs [hunk 1]
```diff
@@ -1,21 +1,10 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System.Collections.Generic;
-using Microsoft.AspNetCore.Http;
-
 namespace Microsoft.AspNetCore.Routing
 {
-    public class LinkGeneratorContext
+    public class LinkOptions
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing.Abstractions/LinkOptions.cs]: I like `LinkGeneratorOptions` more

**Review by JamesNK — APPROVED** [review]
Looks good to me. Run by ryan before checking in

**Review by rynowak — APPROVED** [review]

File: src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs [hunk 2]
```diff
@@ -19,95 +21,213 @@ internal class DefaultLinkGenerator : LinkGenerator
...
+            return TryGetLinkInternal(
+                httpContext,
+                address,
+                explicitValues: values,
+                ambientValues: ambientValues,
+                options,
+                out link);
+        }
+
+        private bool TryGetLinkInternal<TAddress>(
```
[the remark this answers is not in the document]
Comment (kichalla) [reply src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs]: Not all above this line though as the method above this one `TryGetLinkByRouteValues` depends on RouteValues based address which we cannot move to the base type.

File: src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs [hunk 3]
```diff
@@ -1,13 +1,240 @@
...
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (TryGetLink(httpContext, values, options, out var link))
+            {
+                return link;
+            }
+
+            throw new InvalidOperationException("Could not find a matching endpoint to generate a link.");
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs]: Generally if something is optional you don't want to null-check it. It's just not very friendly. You make the user do lots of extra work in the case where something (say the HttpContext) **might** be null in their code.

**Pull request merged**